pkg/controller/controllerutil/controllerutil.go

```diff
@@ -39,10 +41,30 @@ func SetControllerReference(owner, object v1.Object, scheme *runtime.Scheme) err
 		return err
 	}
 
+	// Make sure that object and owner are in the same namespace
+	if object.GetNamespace() != owner.GetNamespace() {
+		return fmt.Errorf("Object %s/%s must be in the same namespace as owner %s/%s", object.GetNamespace(), object.GetName(), owner.GetNamespace(), owner.GetName())
+	}
+
 	// Create a new ref
 	ref := *v1.NewControllerRef(owner, schema.GroupVersionKind{Group: gvk.Group, Version: gvk.Version, Kind: gvk.Kind})
 
-	// Add it to the child
-	object.SetOwnerReferences(append(object.GetOwnerReferences(), ref))
+	existingRefs := object.GetOwnerReferences()
+	fi := -1
+	for i, r := range existingRefs {
+		if r.Kind == gvk.Kind && r.Name == owner.GetName() && r.UID == owner.GetUID() {
+			fi = i
+		} else if r.Controller != nil && *r.Controller {
+			return fmt.Errorf("Object %s/%s is already owned by another %s controller %s", object.GetNamespace(), object.GetName(), r.Kind, r.Name)
```

Comment:
May be we should consider creating an error-type for this error because users would like to detect this error-type and kick-in some business logic for adopting of these objects (re-parenting to new owner). @DirectXMan12 what do you think ?